pkg/controller/mizar/mizar-node-controller.go

```diff
@@ -0,0 +1,293 @@
+	}
+}
+
+// Generate a node key
+func (c *MizarNodeController) genKey(resource *v1.Node) string {
+	nodeKind := resource.GetObjectKind().GroupVersionKind().Kind
+	if nodeKind == "" {
+		nodeKind = NodeKind
+	}
+	nodeTenant := resource.GetTenant()
+	nodeName := resource.GetName()
+	if nodeName == "" {
+		klog.Fatalln(c.logMessage("Node name should not be null", nodeName))
+	}
+	var nodeStatusType, nodeStatus string
+	conditions := resource.Status.Conditions
+	if len(conditions) < 6 {
+		klog.Fatalln(c.logInfoMessage("Node status information is needed"))
+	} else {
+		nodeStatusType = fmt.Sprintf("%s", resource.Status.Conditions[5].Type)
+		nodeStatus = fmt.Sprintf("%s", resource.Status.Conditions[5].Status)
+	}
+	var nodeAddressType, nodeAddress string
+	addresses := resource.Status.Addresses
+	if len(addresses) < 1 {
+		klog.Fatalln(c.logInfoMessage("Node ip address is needed"))
+	} else {
+		nodeAddressType = fmt.Sprintf("%s", resource.Status.Addresses[0].Type)
+		nodeAddress = fmt.Sprintf("%s", resource.Status.Addresses[0].Address)
+	}
+	key := fmt.Sprintf("%s/%s/%s/%s/%s/%s/%s", nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress)
+	return key
+}
+
+// Parse a key and get node information
+func (c *MizarNodeController) parseKey(key string) (nodeKind, nodeTenant, nodeName, nodeStatusType, nodeStatus, nodeAddressType, nodeAddress string, err error) {
+	segs := strings.Split(key, "/")
+	nodeKind = segs[0]
+	nodeTenant = segs[1]
+	nodeName = segs[2]
+	nodeStatusType = segs[3]
+	nodeStatus = segs[4]
+	nodeAddressType = segs[5]
+	nodeAddress = segs[6]
+	if len(segs) < 7 || nodeName == "" {
+		err = fmt.Errorf("Invalid key format - key=%s", key)
+		return
+	}
+	return
+}
+
+func (c *MizarNodeController) determineEventType(key1, key2 string) (eventType EventType, err error) {
+	segs1 := strings.Split(key1, "/")
+	segs2 := strings.Split(key2, "/")
+	if len(segs1) < 7 || len(segs2) < 7 {
+		err = fmt.Errorf("Invalid key format - key1=%s, key2=%s", key1, key2)
+		return
+	}
+	status1 := segs1[4]
+	status2 := segs2[4]
+	eventType = EventType_Update
+	if status1 != status2 && status2 == NodeReady {
+		eventType = EventType_Resume
+	}
+	return
+}
+
+func (c *MizarNodeController) logInfoMessage(info string) string {
+	message := fmt.Sprintf("[NodeController][%s]", info)
+	return message
+}
+
+func (c *MizarNodeController) logMessage(msg string, detail interface{}) string {
+	message := fmt.Sprintf("[NodeController][%s][Node Info]:%v", msg, detail)
+	return message
+}
+
+//gRPC request message, Integration is needed
+func (c *MizarNodeController) gRPCRequest(event EventType, nodeName, nodeAddress string) (response bool, err error) {
+	client, ctx, conn, cancel, err := getGrpcClient(c.grpcHost)
+	if err != nil {
+		klog.Errorf(c.logMessage("gRPC connection failed ", err))
+		return false, err
+	}
+	defer conn.Close()
+	defer cancel()
+	var resource BuiltinsNodeMessage
+	resource = BuiltinsNodeMessage{
```

Comment:
Is there any reason that the existing func ConvertToNodeContract() is not used?

I think the code line 257 - 261 can be rewritten as:
resource := ConvertToNodeContract(lister.Get(node))

Reply:
> Is there any reason that the existing func ConvertToNodeContract() is not used?
> 
> I think the code line 257 - 261 can be rewritten as:
> resource := ConvertToNodeContract(lister.Get(node))

Return data type is different which I need for gRPC communication.